apps/ex_shop/web/gateways/gateway.ex

```diff
@@ -18,6 +18,10 @@ defmodule ExShop.Gateway do
     ExShop.Gateway.BrainTree.authorize(order, payment_method_params["braintree"])
   end
 
+  defp do_authorize_payment(order, "new_payment_gateway", payment_method_params) do
+    ExShop.Billing.Gateway.NewPaymentGateway.authorize(order, payment_method_params["new_payment_gateway"])
```

Comment:
Here we are invoking the payment gateway directly, We start a dedicated worker thread for each payment gateway when the application starts and the gateway can be consumed via a common interface. See: https://github.com/vinsol/ex-shop/blob/umbrella/apps/ex_shop/web/gateways/stripe.ex#L12 for example.